File: examples/node/local.mjs
/**
 * turbolite example — Node.js local compressed
 *
 * A small HTTP API server backed by turbolite-compressed SQLite (local mode).
 * See tiered.mjs for S3 tiered storage.
 *
 * Run:
 *   npm install turbolite
 *   node examples/node/local.mjs
 *
 * Or via Make:
 *   make example-node
 *
 * Then:
 *   curl -X POST localhost:3000/books -d '{"title":"Dune","year":1965}'
 *   curl localhost:3000/books
 */

import { connect } from "turbolite";
import http from "http";
import { mkdtempSync, rmSync } from "fs";
import { tmpdir } from "os";
import { join } from "path";

// ── Set up database ─────────────────────────────────────────────────

const dataDir = mkdtempSync(join(tmpdir(), "turbolite-example-"));
const dbPath = join(dataDir, "books.db");
// File-first local mode (the default): the user-visible artifact is
// `books.db`. Hidden implementation state (manifest, cache, staging logs)
// lives next to it at `books.db-turbolite/`.
//
// `books.db` is turbolite's compressed page image. It is not promised to
// be opened by stock sqlite3. To get a normal SQLite file the standard
// CLI can read, run `db.exec("VACUUM INTO 'books-export.sqlite'")`.
const db = connect(dbPath);

db.exec(`
  CREATE TABLE books (
    id INTEGER PRIMARY KEY AUTOINCREMENT,
    title TEXT NOT NULL,
    year INTEGER
  )
`);

// ── HTTP server ─────────────────────────────────────────────────────

const listAll = db.prepare("SELECT * FROM books ORDER BY year");
const insertBook = db.prepare("INSERT INTO books (title, year) VALUES (?, ?)");
const lastBook = db.prepare("SELECT * FROM books ORDER BY id DESC LIMIT 1");

const server = http.createServer((req, res) => {
  if (req.method === "GET" && req.url === "/books") {
    res.writeHead(200, { "Content-Type": "application/json" });
    res.end(JSON.stringify(listAll.all()));

  } else if (req.method === "POST" && req.url === "/books") {
    let body = "";
    req.on("data", (c) => (body += c));
    req.on("end", () => {
      const { title, year } = JSON.parse(body);
      insertBook.run(title, year);
      res.writeHead(201, { "Content-Type": "application/json" });
      res.end(JSON.stringify(lastBook.get()));
    });

  } else {
    res.writeHead(404);
    res.end("Not found");
  }
});

process.on("SIGINT", () => {
  db.close();
  rmSync(dataDir, { recursive: true, force: true });
  process.exit(0);
});

server.listen(3000, () => {
  console.log(`turbolite — listening on :3000`);
});

File: examples/c/local.c
/*
 * turbolite example — C local compressed
 *
 * A sensor data logger that writes readings to compressed SQLite (local mode).
 * See tiered.c for S3 tiered storage.
 *
 * Build & run:
 *   make lib-bundled header
 *   make example-c
 */

#include "../../turbolite.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

static void die(const char *msg) {
    const char *err = turbolite_last_error();
    fprintf(stderr, "FATAL: %s: %s\n", msg, err ? err : "unknown");
    exit(1);
}

static char tmpdir_buf[256];

int main(void) {
    /* Create data directory */
    snprintf(tmpdir_buf, sizeof(tmpdir_buf), "%s/turbolite-sensor-XXXXXX",
             getenv("TMPDIR") ? getenv("TMPDIR") : "/tmp");
    const char *data_dir = mkdtemp(tmpdir_buf);
    if (!data_dir) { perror("mkdtemp"); return 1; }

    printf("turbolite %s — sensor logger\n", turbolite_version());
    printf("Data dir: %s\n\n", data_dir);

    /*
     * File-first registration: the caller's database path is the local
     * page image. Sidecar metadata lives at `<db_path>-turbolite/`.
     *
     * To export a stock SQLite file from this database, run:
     *   turbolite_exec(db, "VACUUM INTO '/tmp/exported.sqlite'");
     * The exported file opens with the standard sqlite3 CLI.
     */
    char db_path[512];
    snprintf(db_path, sizeof(db_path), "%s/sensors.db", data_dir);

    if (turbolite_register_local_file_first("sensor", db_path, 3) != 0)
        die("register VFS");

    void *db = turbolite_open(db_path, "sensor");
    if (!db) die("open database");

    turbolite_exec(db,
        "CREATE TABLE readings ("
        "  id INTEGER PRIMARY KEY AUTOINCREMENT,"
        "  sensor TEXT NOT NULL,"
        "  value REAL NOT NULL,"
        "  ts INTEGER NOT NULL"
        ")");

    /* Simulate logging sensor readings */
    srand((unsigned)time(NULL));
    for (int i = 0; i < 10; i++) {
        char sql[256];
        double temp = 20.0 + (rand() % 100) / 10.0;
        snprintf(sql, sizeof(sql),
            "INSERT INTO readings (sensor, value, ts) VALUES ('temp-01', %.1f, %ld)",
            temp, (long)time(NULL) + i);

        if (turbolite_exec(db, sql) != 0)
            die("insert reading");
    }
    printf("Logged 10 readings.\n");

    /* Query summary */
    char *json = turbolite_query_json(db,
        "SELECT sensor, COUNT(*) as count, "
        "ROUND(AVG(value), 1) as avg_temp, "
        "ROUND(MIN(value), 1) as min_temp, "
        "ROUND(MAX(value), 1) as max_temp "
        "FROM readings GROUP BY sensor");
    if (!json) die("query");

    printf("Summary: %s\n", json);
    turbolite_free_string(json);

    turbolite_close(db);
    return 0;
}
